gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java

```diff
@@ -35,6 +38,33 @@
     Page<T> nextPage();
   }
 
+  static class PageIterator<T> extends AbstractIterator<T> {
+
+    private Iterator<T> currentPageIterator;
+    private Page<T> currentPage;
+
+    PageIterator(Page<T> currentPage) {
+      this.currentPageIterator = currentPage.values().iterator();
+      this.currentPage = currentPage;
+    }
+
+    @Override
+    protected T computeNext() {
+      if (currentPageIterator.hasNext()) {
+        return currentPageIterator.next();
+      }
+      Page<T> nextPage = currentPage.nextPage();
+      if (nextPage != null) {
+        currentPage = nextPage;
+        currentPageIterator = currentPage.values().iterator();
```

Comment:
"Once the implementation either invokes `endOfData` or throws an exception, `computeNext` is guaranteed to never be called again."

So,  I think it should be safe to assign `currentPage.nextPage()` directly to currentPage.

Also, we should change the `if` to a `while` to support the case of an "empty" nextPage that is not last (which theoretically may happen). Maybe something like:
```java
while (!currentPageIterator.hasNext()) {
  currentPage = currentPage.nextPage();
  if (currentPage == null) {
    return endOfData();
  }
  currentPageIterator = currentPage.values().iterator();
}
return currentPageIterator.next();
```